File: android/network/src/main/kotlin/com/ridelink/network/security/DeviceIdentity.kt
package com.ridelink.network.security

import com.ridelink.core.model.SpkiHash
import com.ridelink.core.security.IdentityCertificate
import com.ridelink.core.security.UtcTime
import java.io.ByteArrayInputStream
import java.security.KeyStore
import java.security.PublicKey
import java.security.SecureRandom
import java.security.Signature
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.security.interfaces.ECPublicKey
import javax.net.ssl.KeyManager
import javax.net.ssl.KeyManagerFactory

/**
 * The device's long-term RideLink identity, in the shape the TLS layer needs it (ADR-012, ADR-017).
 *
 * The private key is **never** here — it stays behind an opaque `PrivateKey` handle whose material
 * lives in Android Keystore. [keyManagers] hands that handle to JSSE, which is the only thing that
 * ever uses it, and it is never serialised, never written to a file and never logged.
 */
class DeviceIdentity internal constructor(
    val certificate: X509Certificate,
    private val keyStore: KeyStore,
    private val alias: String,
    private val keyPassword: CharArray?,
) {
    /**
     * `identity_spki_sha256`. Computed from `PublicKey.getEncoded()`, which on the JVM/Android
     * **is** the DER SubjectPublicKeyInfo — so this is a hash of the exact bytes, not a
     * reconstruction (ADR-017 §4).
     */
    val identitySpkiSha256: SpkiHash = IdentityCertificate.spkiHashOfEncoded(certificate.publicKey.encoded)

    fun keyManagers(): Array<KeyManager> {
        val factory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm())
        factory.init(keyStore, keyPassword)
        return factory.keyManagers
    }

    companion object {
        /** RideLink's key/signature pair, fixed for both platforms by ADR-017 §1. */
        const val KEY_ALGORITHM: String = "EC"
        const val CURVE: String = "secp256r1"
        const val SIGNATURE_ALGORITHM: String = "SHA256withECDSA"
    }
}

/**
 * Turns a P-256 keypair into the RideLink identity certificate, using
 * [com.ridelink.core.security.IdentityCertificate] for the encoding and the platform's own
 * `Signature` for the signing (ADR-017 §3 — we arrange bytes, the platform does the cryptography).
 *
 * Kept free of `android.*` on purpose: the encoding, the signing call and the extraction below are
 * then all exercisable by a plain JVM unit test against a real X.509 parser, instead of only on a
 * phone. [AndroidKeystoreIdentityStore] is the thin part that is not.
 */
object IdentityIssuer {
    /**
     * Issues (or re-issues) a self-signed certificate around [publicKey].
     *
     * Re-issuance around an **unchanged** key is a first-class operation, not an afterthought:
     * ADR-012 says the pin is the key, so a fresh certificate for an old key is a non-event that
     * must not prompt the user. It is also the reason RideLink does not use Android's
     * `KeyGenParameterSpec` auto-issued certificate, which can only be minted when the key is
     * created (ADR-017 §3).
     *
     * @param signer already initialised for signing with the matching private key.
     * @param issuedAt wall-clock time, supplied by the caller. X.509 validity is defined in
     *   wall-clock terms by RFC 5280 and has no monotonic alternative — see
     *   [com.ridelink.core.security.UtcTime] for why this is the one permitted exception.
     */
    fun issue(
        publicKey: PublicKey,
        signer: Signature,
        issuedAt: UtcTime,
        random: SecureRandom = SecureRandom(),
    ): X509Certificate {
        val point = uncompressedPoint(publicKey)
        val (notBefore, notAfter) = IdentityCertificate.validityWindow(issuedAt)
        val tbs = IdentityCertificate.tbsCertificate(point, freshSerial(random), notBefore, notAfter)
        signer.update(tbs)
        val der = IdentityCertificate.certificate(tbs, signer.sign())
        return parseCertificate(der)
    }

    /**
     * 16 CSPRNG bytes with the top bit cleared, so the DER INTEGER is unambiguously positive
     * without a leading pad byte (ADR-017 §2).
     */
    fun freshSerial(random: SecureRandom = SecureRandom()): ByteArray {
        val serial = ByteArray(IdentityCertificate.SERIAL_BYTES)
        random.nextBytes(serial)
        serial[0] = (serial[0].toInt() and 0x7F).toByte()
        return serial
    }

    /** Round-trips our own DER through the platform's X.509 parser, so a malformed encoding fails here. */
    fun parseCertificate(der: ByteArray): X509Certificate =
        CertificateFactory.getInstance("X.509")
            .generateCertificate(ByteArrayInputStream(der)) as X509Certificate

    /**
     * The raw uncompressed X9.63 point (`0x04 ‖ X(32) ‖ Y(32)`) for a P-256 public key.
     *
     * The curve check is done by *rebuilding* the SubjectPublicKeyInfo from the extracted point
     * and requiring it to equal `publicKey.encoded`. That is stronger than inspecting curve
     * parameters and cheaper to be sure of: it rejects every non-P-256 key, and it simultaneously
     * proves at runtime that the two routes to `identity_spki_sha256` — Android's
     * hash-the-encoded-SPKI and iOS's rebuild-from-the-point (ADR-017 §4) — agree on this key.
     */
    fun uncompressedPoint(publicKey: PublicKey): ByteArray {
        require(publicKey is ECPublicKey) { "RideLink identity keys are EC P-256 (ADR-017)" }
        val point =
            byteArrayOf(UNCOMPRESSED_TAG) +
                fixedWidth(publicKey.w.affineX.toByteArray()) +
                fixedWidth(publicKey.w.affineY.toByteArray())
        require(IdentityCertificate.subjectPublicKeyInfo(point).contentEquals(publicKey.encoded)) {
            "public key is not P-256 in the canonical SubjectPublicKeyInfo encoding (ADR-017 §1)"
        }
        return point
    }

    private const val UNCOMPRESSED_TAG: Byte = 0x04
    private const val COORDINATE_BYTES = 32

    /**
     * `BigInteger.toByteArray()` is signed and variable width: it drops leading zero bytes and
     * prepends one when the top bit is set. A P-256 coordinate must be exactly 32 bytes, so both
     * cases have to be corrected — silently getting this wrong shifts the whole point and produces
     * a valid-looking but different `identity_spki_sha256`.
     */
    private fun fixedWidth(signedMagnitude: ByteArray): ByteArray {
        val stripped =
            if (signedMagnitude.size > COORDINATE_BYTES) {
                require(signedMagnitude.size == COORDINATE_BYTES + 1 && signedMagnitude[0] == 0x00.toByte()) {
                    "EC coordinate wider than $COORDINATE_BYTES bytes"
                }
                signedMagnitude.copyOfRange(1, signedMagnitude.size)
            } else {
                signedMagnitude
            }
        if (stripped.size == COORDINATE_BYTES) return stripped
        return ByteArray(COORDINATE_BYTES - stripped.size) + stripped
    }
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Security/DeviceIdentity.swift
import CryptoKit
import Foundation
import RideLinkCore
import Security

/// The device's long-term RideLink identity, in the shape the TLS layer needs it
/// (ADR-012, ADR-017).
///
/// The private key never appears here as bytes. It stays behind an opaque `SecKey` handle whose
/// material lives in the Keychain, and the only operation performed with it is
/// `SecKeyCreateSignature` — there is no export, no serialisation and no log path.
public struct DeviceIdentity: @unchecked Sendable {
    /// `identity_spki_sha256`, the only pinned value in the system (ADR-012).
    public let identitySpkiSha256: SpkiHash

    /// The DER certificate this device presents during the TLS handshake.
    public let certificateDER: Data

    /// What `Network.framework` needs as a local TLS identity.
    public let secIdentity: SecIdentity

    /// The raw uncompressed X9.63 point of the identity public key.
    public let publicKeyPoint: [UInt8]
}

public enum DeviceIdentityError: Error, CustomStringConvertible {
    case keyGenerationFailed(String)
    case publicKeyUnavailable(String)
    case signingFailed(String)
    case certificateRejectedByParser
    case identityCreationFailed
    case keychainFailure(OSStatus)
    case unsupportedKey(String)

    public var description: String {
        switch self {
        case let .keyGenerationFailed(reason): return "identity key generation failed: \(reason)"
        case let .publicKeyUnavailable(reason): return "identity public key unavailable: \(reason)"
        case let .signingFailed(reason): return "certificate signing failed: \(reason)"
        case .certificateRejectedByParser: return "SecCertificateCreateWithData rejected the encoded certificate"
        case .identityCreationFailed: return "SecIdentityCreate returned nil"
        case let .keychainFailure(status): return "keychain operation failed with OSStatus \(status)"
        case let .unsupportedKey(reason): return "unsupported identity key: \(reason)"
        }
    }
}

/// Creates and loads the device identity (ADR-017 §1: P-256, `ecdsa-with-SHA256`, one algorithm
/// for both platforms).
///
/// **Where the key lives.** `.keychain` stores a *permanent* key under an application tag, which
/// is what the app uses. `.ephemeral` produces a transient key that vanishes with the process; it
/// exists because an unsigned test binary has no keychain entitlement, so it is the only way the
/// certificate encoding, `SecIdentityCreate` and the TLS handshake can be exercised by
/// `swift test` on a laptop. It is named for what it is and never selected by the app.
///
/// **Lifetime of the keychain key, by platform, stated because it differs and users notice:**
///
/// | Event | Key survives? |
/// |---|---|
/// | App restart / device reboot | Yes |
/// | App upgrade (same signing identity, same keychain access group) | Yes |
/// | Uninstall / reinstall | **Usually yes on iOS** — unlike Android, iOS has historically left
///   keychain items behind after an app is deleted. Not relied upon: a reinstall that *does* lose
///   it simply produces a new identity, which is a `pin_mismatch` on the peer and is recovered by
///   forget-and-re-pair, exactly as ADR-012 specifies |
/// | "Reset all content and settings" | No |
///
/// A different key *is* a different identity (ADR-012), so losing it is never silently papered
/// over — it produces a security warning the user resolves deliberately.
public struct DeviceIdentityStore {
    public enum Storage: Sendable {
        /// Production: a permanent Keychain key under `applicationTag`.
        case keychain(applicationTag: String)
        /// **Test only.** A transient key that needs no keychain entitlement and does not persist.
        case ephemeral
    }

    public static let defaultApplicationTag = "com.ridelink.identity.v1"

    private let storage: Storage

    public init(storage: Storage = .keychain(applicationTag: defaultApplicationTag)) {
        self.storage = storage
    }

    /// Returns the existing identity, or creates one on first run.
    ///
    /// If a stored certificate has fallen outside its validity window, a fresh one is issued
    /// **around the same key**: the SPKI is unchanged, so peers stay trusted and no six-digit code
    /// is shown (ADR-012, PROTOCOL §4.5.3).
    ///
    /// - Parameter now: wall-clock time, supplied by the caller. X.509 validity is defined in
    ///   wall-clock terms by RFC 5280 and has no monotonic alternative — see `UtcTime` for why
    ///   this is the one permitted exception to CLAUDE.md's monotonic-clocks rule.
    public func loadOrCreate(now: UtcTime) throws -> DeviceIdentity {
        let privateKey: SecKey
        switch storage {
        case let .keychain(applicationTag):
            privateKey = try loadKeychainKey(applicationTag: applicationTag) ?? generateKeychainKey(applicationTag: applicationTag)
        case .ephemeral:
            privateKey = try generateKey(permanent: false, applicationTag: nil)
        }
        return try issueIdentity(privateKey: privateKey, now: now)
    }

    /// Builds the certificate and the `SecIdentity` around an existing key. Re-issuing is a
    /// first-class operation, not an afterthought: ADR-012 pins the key, so a fresh certificate
    /// for an old key must be a non-event.
    public func issueIdentity(privateKey: SecKey, now: UtcTime) throws -> DeviceIdentity {
        let point = try uncompressedPoint(of: privateKey)
        let (notBefore, notAfter) = IdentityCertificate.validityWindow(issuedAt: now)
        let tbs = try IdentityCertificate.tbsCertificate(
            uncompressedPoint: point,
            serial: Self.freshSerial(),
            notBefore: notBefore,
            notAfter: notAfter
        )

        var error: Unmanaged<CFError>?
        guard let signature = SecKeyCreateSignature(
            privateKey, .ecdsaSignatureMessageX962SHA256, Data(tbs) as CFData, &error
        ) as Data? else {
            throw DeviceIdentityError.signingFailed("\(error?.takeRetainedValue().localizedDescription ?? "unknown")")
        }

        let der = Data(try IdentityCertificate.certificate(tbsCertificate: tbs, signature: [UInt8](signature)))
        // Round-tripping through Apple's own parser here means a malformed encoding fails at
        // creation, on this device, rather than as an unexplained handshake failure on the peer.
        guard let certificate = SecCertificateCreateWithData(nil, der as CFData) else {
            throw DeviceIdentityError.certificateRejectedByParser
        }
        // SecIdentityCreate, NOT SecIdentityCreateWithCertificate: the latter is macOS-only
        // (`SEC_OS_OSX` / `__IPHONE_NA`) and is the function most iOS examples reach for by
        // mistake. This one takes the key directly and needs no PKCS#12 round trip, so the private
        // key is never exported (ADR-017 §5).
        guard let identity = SecIdentityCreate(nil, certificate, privateKey) else {
            throw DeviceIdentityError.identityCreationFailed
        }

        return DeviceIdentity(
            identitySpkiSha256: try IdentityCertificate.identitySpkiSha256(uncompressedPoint: point),
            certificateDER: der,
            secIdentity: identity,
            publicKeyPoint: point
        )
    }

    /// 16 CSPRNG bytes with the top bit cleared, so the DER INTEGER is unambiguously positive
    /// without a leading pad byte (ADR-017 §2).
    public static func freshSerial() -> [UInt8] {
        var serial = [UInt8](repeating: 0, count: IdentityCertificate.serialBytes)
        _ = SecRandomCopyBytes(kSecRandomDefault, serial.count, &serial)
        serial[0] &= 0x7F
        return serial
    }

    // MARK: - Key material

    private func generateKey(permanent: Bool, applicationTag: String?) throws -> SecKey {
        var attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits as String: 256,
        ]
        if permanent, let applicationTag {
            attributes[kSecPrivateKeyAttrs as String] = [
                kSecAttrIsPermanent as String: true,
                kSecAttrApplicationTag as String: Data(applicationTag.utf8),
                // The identity must be usable while the screen is locked, which is the normal
                // state of a phone in a tank bag mid-ride (ARCHITECTURE §6.4). Hence
                // AfterFirstUnlock rather than WhenUnlocked, and no biometric gate.
                kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            ]
        } else {
            attributes[kSecPrivateKeyAttrs as String] = [kSecAttrIsPermanent as String: false]
        }

        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
            throw DeviceIdentityError.keyGenerationFailed("\(error?.takeRetainedValue().localizedDescription ?? "unknown")")
        }
        return key
    }

    private func generateKeychainKey(applicationTag: String) throws -> SecKey {
        try generateKey(permanent: true, applicationTag: applicationTag)
    }

    private func loadKeychainKey(applicationTag: String) throws -> SecKey? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrApplicationTag as String: Data(applicationTag.utf8),
            kSecReturnRef as String: true,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        switch status {
        case errSecSuccess:
            guard let item else { return nil }
            // A CFTypeRef from a kSecReturnRef query on kSecClassKey is a SecKey.
            return (item as! SecKey) // swiftlint:disable:this force_cast
        case errSecItemNotFound:
            return nil
        default:
            throw DeviceIdentityError.keychainFailure(status)
        }
    }

    /// The raw uncompressed X9.63 point for a P-256 key.
    ///
    /// `SecKeyCopyExternalRepresentation` returns the bare point rather than a
    /// SubjectPublicKeyInfo, which is the asymmetry ADR-017 §4 records: Android reads an SPKI
    /// straight out of `PublicKey.getEncoded()`, iOS has to rebuild one. The length and leading
    /// `0x04` are checked by `IdentityCertificate`, so a non-P-256 key can never reach the hash
    /// and become somebody's pinned identity.
    private func uncompressedPoint(of privateKey: SecKey) throws -> [UInt8] {
        guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
            throw DeviceIdentityError.publicKeyUnavailable("SecKeyCopyPublicKey returned nil")
        }
        var error: Unmanaged<CFError>?
        guard let raw = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
            throw DeviceIdentityError.publicKeyUnavailable(
                "\(error?.takeRetainedValue().localizedDescription ?? "unknown")")
        }
        return [UInt8](raw)
    }
}

/// Certificate facts RideLink needs from a peer, derived with Apple's own X.509 implementation
/// rather than by parsing DER ourselves.
public enum PeerCertificateInspector {
    /// `identity_spki_sha256` for a peer certificate (ADR-012).
    ///
    /// Rebuilt from the peer's public key, and rejected unless it is a P-256 key in the canonical
    /// encoding — so a peer offering RSA or a different curve is refused before its key can be
    /// hashed into something that looks like an identity.
    public static func identitySpkiSha256(of certificate: SecCertificate) throws -> SpkiHash {
        guard let key = SecCertificateCopyKey(certificate) else {
            throw DeviceIdentityError.publicKeyUnavailable("SecCertificateCopyKey returned nil")
        }
        guard let attributes = SecKeyCopyAttributes(key) as? [String: Any],
              let keyType = attributes[kSecAttrKeyType as String] as? String,
              keyType == (kSecAttrKeyTypeECSECPrimeRandom as String),
              let bits = attributes[kSecAttrKeySizeInBits as String] as? Int, bits == 256
        else {
            throw DeviceIdentityError.unsupportedKey("peer key is not P-256 (ADR-017 §1)")
        }
        var error: Unmanaged<CFError>?
        guard let raw = SecKeyCopyExternalRepresentation(key, &error) as Data? else {
            throw DeviceIdentityError.publicKeyUnavailable(
                "\(error?.takeRetainedValue().localizedDescription ?? "unknown")")
        }
        return try IdentityCertificate.identitySpkiSha256(uncompressedPoint: [UInt8](raw))
    }

    /// The structural half of PROTOCOL §4.1's certificate check: well-formed, self-signature
    /// verifies, inside its validity window.
    ///
    /// Implemented by evaluating the certificate **against itself as the anchor** with a basic
    /// X.509 policy, which is exactly "is this a valid self-signed certificate right now" and
    /// deliberately *not* a chain or hostname check — RideLink has no CA and no name to check
    /// (ARCHITECTURE §4.3). A false here becomes `ERROR/certificate_invalid`, a distinct code from
    /// `pin_mismatch` so a phone with a wrong clock is not reported to its owner as an attack.
    public static func isStructurallyValid(_ certificate: SecCertificate) -> Bool {
        var trust: SecTrust?
        let policy = SecPolicyCreateBasicX509()
        guard SecTrustCreateWithCertificates(certificate, policy, &trust) == errSecSuccess,
              let trust
        else {
            return false
        }
        guard SecTrustSetAnchorCertificates(trust, [certificate] as CFArray) == errSecSuccess,
              SecTrustSetAnchorCertificatesOnly(trust, true) == errSecSuccess
        else {
            return false
        }
        return SecTrustEvaluateWithError(trust, nil)
    }
}
